### PR TITLE
[SP-93] Added LogOut functionality

### PR DESCRIPTION
**Pull request by innowate**

### CODE REVIEW

**Comment by Ne4istb** [comment]
@innowate Please, fix CI 🔨

File: lib/account-management/auth.action-creator.dart [hunk 1]
```diff
@@ -1,17 +1,30 @@
...
+          'password': credentials['password'],
+          'rememberMe': credentials['rememberMe']
+        });
   }
 
   static Action receiveLogin() {
     return new Action(LOGIN_SUCCESS);
   }
 
+  static Action logOut() {
```
Comment (Ne4istb) [inline lib/account-management/auth.action-creator.dart]: ✏️ 'Logout' is single world so no camel case is needed. Please, fix it in other places as well.

File: lib/account-management/auth.reducer.dart [hunk 2]
```diff
@@ -28,6 +28,11 @@ class AuthReducer {
           ..['isFetching'] = true
           ..['isAuthenticated'] = false;
 
+      case LOGOUT_REQUEST:
+        return new State.from(state)
+          ..['isFetching'] = true
+          ..['isAuthenticated'] = false;
```
Comment (Ne4istb) [inline lib/account-management/auth.reducer.dart]: isAuthenticated should be changed on LOGOUT_SUCCESS action.

File: lib/account-management/log-out/log-out.component.dart [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import 'package:SmartPeopleUI/index.dart' show AuthActionCreator, DialogAction;
+import 'package:SmartPeopleUI/shared/components/controls/dialog/dialog.component.dart';
+
+@Component(
+    selector: 'sp-log-out',
+    templateUrl: 'log-out.component.html',
+    directives: const[DialogComponent]
+)
+
+class ConfirmationToLogOut implements OnInit{
```
Comment (Ne4istb) [inline lib/account-management/log-out/log-out.component.dart]: ✏️ Follow Angular2 official style guide in naming convention (https://angular.io/docs/ts/latest/guide/style-guide.html#02-03)

File: lib/account-management/log-out/log-out.component.dart [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+   ConfirmationToLogOut(this._store, this._router);
+
+   @override
+   ngOnInit() {
+      dialogActions.add(new DialogAction('Yes', _onYesClick));
+      dialogActions.add(new DialogAction('No', _onNoClick));
+   }
+
+   onLogOut(){
```
Comment (Ne4istb) [inline lib/account-management/log-out/log-out.component.dart]: ✏️ public component method name should show the intention. Rename to logout().

File: lib/shared/middleware/api.middleware.dart [hunk 5]
```diff
@@ -61,6 +62,19 @@ class ApiMiddleware {
     }
   }
 
+  Future<Action> _logOut(Action action) async {
+    try {
+
```
Comment (Ne4istb) [inline lib/shared/middleware/api.middleware.dart]: Here should be a request to server which says that user is trying to logout.

File: lib/shared/middleware/api.middleware.dart [hunk 6]
```diff
@@ -61,6 +62,19 @@ class ApiMiddleware {
     }
   }
 
+  Future<Action> _logOut(Action action) async {
+    try {
+
+        _localStorage.clear();
```
Comment (Ne4istb) [inline lib/shared/middleware/api.middleware.dart]: Don't clear storages! There could be another important info. Remove only records associated with token.

File: lib/account-management/auth.reducer.dart (hunk 2 above)
Comment (innowate) [inline lib/account-management/auth.reducer.dart]: may be isAuthenticated should be changed on LOGOUT_REQUEST action?

Comment (Ne4istb) [inline lib/account-management/auth.reducer.dart]: No, by workflow logout request is just a request, which do not guarantee a logout. State should be changed when the operation is succeeded.

File: client/lib/account-management/logout/logout.component.dart [hunk 7]
```diff
@@ -25,21 +25,21 @@ class LogoutComponent implements OnInit{
...
+      dialogActions.add(new DialogAction('Yes', onYesClick));
+      dialogActions.add(new DialogAction('No', onNoClick));
    }
 
    logout(){
       dialog.showModal();
    }
 
-   _onYesClick() {
+   onYesClick() {
```
Comment (Ne4istb) [inline client/lib/account-management/logout/logout.component.dart]: It should not be public

File: client/lib/account-management/logout/logout.component.dart [hunk 8]
```diff
@@ -25,21 +25,21 @@ class LogoutComponent implements OnInit{
...
       _store.dispatch(AuthActionCreator.requestLogout());
       _router.navigate(['Login']);
       dialog.close();
    }
 
-   _onNoClick() {
+   onNoClick() {
```
Comment (Ne4istb) [inline client/lib/account-management/logout/logout.component.dart]: It should not be public

File: client/test/account-management/logout/logout.component.dart [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      setUp(() {
+        component = new LogoutComponent(mockStore, mockRouter);
+
+        mockDialog = spy(new MockDialogComponent(), new DialogComponent());
+        when(mockDialog.showModal()).thenReturn({});
+        when(mockDialog.close()).thenReturn({});
+        component.dialog = mockDialog;
+      });
+
+      group('On state change', () {
```
Comment (Ne4istb) [inline client/test/account-management/logout/logout.component.dart]: We do not have changing state functionality in logout component. probably another group name?

File: client/test/account-management/logout/logout.component.dart [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        test('Should request logout', () {
+          component.onYesClick();
+
+          var isRequestLogoutAction = predicate((action) => action.type == LOGOUT_REQUEST);
+
+          expect(verify(mockStore.dispatch(argThat(isRequestLogoutAction))).callCount, 1);
+        });
+
+        test('Should return logout action', () {
+          Action result = AuthActionCreator.logoutSuccess();
```
Comment (Ne4istb) [inline client/test/account-management/logout/logout.component.dart]: Action creator tests are in other file.

File: client/test/account-management/logout/logout.component.dart [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+          var logoutAction = component.dialogActions[0];
+          logoutAction.execute();
+        });
+
+          test('Should go to login page after logout action', () {
+            verify(mockRouter.navigate(['Login']));
+          });
+
+          test('Should close dialog after confirm action', () async {
+            var confirmAction = component.dialogActions[0];
```
Comment (Ne4istb) [inline client/test/account-management/logout/logout.component.dart]: Messing up different tests. In setUp method you called component.dialogActions[0] (on yes click). so the group should test only it. confirmAction.execute() Should not be called twice during the test.
Cancel functionality should be in other group.

Comment (Ne4istb) [inline client/test/account-management/logout/logout.component.dart]: Also check action dispatching in this group.

File: client/test/shared/middleware/api.middleware.dart [hunk 12]
```diff
@@ -317,6 +317,31 @@ class ApiMiddlewareTests {
           expect(result.data, null);
         });
       });
+
+      group('Logout request action', () {
+        var action = new Action(LOGOUT_REQUEST);
+        test('Should clear storages if requested logout', () async {
+
+          Response fakeResponse = new Response('{"token": "some_cool_token"}', 200);
```
Comment (Ne4istb) [inline client/test/shared/middleware/api.middleware.dart]: in this case response body is just '{}'

File: client/test/shared/middleware/api.middleware.dart [hunk 13]
```diff
@@ -317,6 +317,31 @@ class ApiMiddlewareTests {
...
+
+          verify(sessionStorage.remove(ApiMiddleware.TOKEN_KEY));
+          verify(localStorage.remove(ApiMiddleware.TOKEN_KEY));
+
+        });
+
+        test('Should try to unauthorize if requested logout', () async {
+          Response fakeResponse = new Response('{"token": "some_cool_token"}', 200);
+
+          when(httpClient.get(argThat(anything), headers: argThat(anything))).thenReturn(fakeResponse);
```
Comment (Ne4istb) [inline client/test/shared/middleware/api.middleware.dart]: Also check if token was passed in authorization header. It's important for the logout.

File: client/test/account-management/auth.reducer.dart [hunk 14]
```diff
@@ -28,6 +28,10 @@ class AuthReducerTests {
           'action': new Action(LOGOUT_SUCCESS),
           'result': new State({'isFetching': true, 'isAuthenticated': false})
```
Comment (Ne4istb) [inline client/test/account-management/auth.reducer.dart]: isFetching should be false here

File: client/test/account-management/logout/logout.component.dart [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+import '../../helpers/angular.dart' ;
+import '../../helpers/mocks.dart' as mocks;
+import '../../helpers/mocks.dart';
+
+class MockDialogComponent extends Mock implements DialogComponent {
+  noSuchMethod(i) => super.noSuchMethod(i);
+}
+
+class LogoutComponentTests {
```
Comment (Ne4istb) [inline client/test/account-management/logout/logout.component.dart]: Also check if logout() method

File: client/test/shared/middleware/api.middleware.dart [hunk 16]
```diff
@@ -333,14 +334,55 @@ class ApiMiddlewareTests {
...
+          Response fakeResponse = new Response('{}', 200);
 
           when(httpClient.get(argThat(anything), headers: argThat(anything))).thenReturn(fakeResponse);
 
           var result = await middleware.apply(store)(next)(action);
 
           expect(result.type, LOGOUT_SUCCESS);
         });
+
+        test('Should check if token was passed in authorization header', () async {
```
Comment (innowate) [inline client/test/shared/middleware/api.middleware.dart]: three tests in api.middleware are not working((

**Pull request merged**